ChatRPG/ChatRPG/Services/EfPersisterService.cs

```diff
@@ -0,0 +1,58 @@
+﻿using ChatRPG.Data;
+using ChatRPG.Data.Models;
+using Microsoft.EntityFrameworkCore;
+
+namespace ChatRPG.Services;
+
+/// <summary>
+/// Service for persisting and loading changes from the data model using Entity Framework.
+/// </summary>
+public class EfPersisterService : IPersisterService
+{
+    private readonly ILogger<EfPersisterService> _logger;
+    private readonly ApplicationDbContext _dbContext;
+
+    public EfPersisterService(ILogger<EfPersisterService> logger, ApplicationDbContext dbContext)
+    {
+        _logger = logger;
+        _dbContext = dbContext;
+    }
+
+    /// <inheritdoc />
+    public async Task SaveAsync(Campaign campaign)
+    {
+        await using var transaction = await _dbContext.Database.BeginTransactionAsync();
+        try
+        {
+            if (!_dbContext.Campaigns.Contains(campaign))
+            {
+                _dbContext.Campaigns.Add(campaign);
+            }
+
+            await _dbContext.SaveChangesAsync();
+            await transaction.CommitAsync();
+        }
+        catch (Exception e)
+        {
+            _logger.LogError(e, "An error occurred while saving");
+            await transaction.RollbackAsync();
+            throw;
+        }
+    }
+
+    /// <inheritdoc />
+    public async Task<Campaign> LoadFromCampaignIdAsync(int campaignId)
+    {
+        return await _dbContext.Campaigns
+            .Where(campaign => campaign.Id == campaignId)
+            .Include(campaign => campaign.Messages)
+            .Include(campaign => campaign.StartScenario)
+            .Include(campaign => campaign.Environments)
+            .Include(campaign => campaign.Events)
+            .Include(campaign => campaign.Characters)
+                .ThenInclude(character => character.CharacterAbilities)
+                .ThenInclude(characterAbility => characterAbility.Ability)
```

Comment:
Er det sådan man definerer eager loading?

Reply:
Jep, og lazy loading er default. 
Så for eksempel linje 52-54:
- `Include(campaign => campaign.Characters)` loader alle denne campaigns characters
- `ThenInclude(character => character.CharacterAbilities)` er kædet sammen med ovenstående, og loader alle `CharacterAbility` som er på characters i denne campaign  
- `ThenInclude(characterAbility => characterAbility.Ability)` fortsætter kæden og loader alle `Ability` som de loadede `CharacterAbility` ovenfor peger på.

Reply:
Det tænkte jeg nok, tak for uddybningen!